### PR TITLE
Time tracking

### PR DESCRIPTION
**Pull request by za3k**
Not actually ready to merge; DEMO for Jeff

### CODE REVIEW

**Comment by WuTheFWasThat** [comment]
can these last 2 commits actually go in a separate branch?

**Comment by za3k** [comment]
Let me finish the enable/disable menu today and then I'll make a PR for that.

**Comment by za3k** [comment]
Jeff, some of this should get merged, d2589ae40b15e7ba0ecc23f8ce4aeb1c30277fac thru bc6d5154cc66caf5da371deddcf1f30f1d6c8c5f. I couldn't figure out how to separate that into its own branch.

**Comment by WuTheFWasThat** [comment]
rebasing this branch

File: assets/js/view.coffee [hunk 1]
```diff
@@ -1230,6 +1235,16 @@ window?.renderLine = renderLine
 
       lineContents = renderLine lineData, lineoptions
       [].push.apply results, lineContents
+      lineContents = @applyRenderHook 'lineContents', lineContents, { row: row }
```
Comment (WuTheFWasThat) [inline assets/js/view.coffee]: this should be before the push.apply

File: assets/js/data.coffee [hunk 2]
```diff
@@ -369,42 +378,63 @@ class Data extends EventEmitter
...
     original_ancestry = @allAncestors child_id, { inclusive: true }
+    new_ancestry = _.union (@allAncestors child_id, { inclusive: true }), (@allAncestors parent_id, { inclusive: true })
+    added_ancestry = _.difference new_ancestry, original_ancestry
+    isFirst = (@_getParents child_id).length == 0
+
+    # Notify all ancestors thaat their list of descendants changed
+    for ancestorId in added_ancestry
+      @emit "afterDescendantAdded", { ancestorId: ancestorId, descendantId: child_id }
+    if isFirst
+      @emit "afterRowAdded", { id: child_id }
```
Comment (WuTheFWasThat) [inline assets/js/data.coffee]: these should be before*

File: assets/js/data.coffee [hunk 3]
```diff
@@ -305,6 +306,14 @@ class Data extends EventEmitter
           visit parent
     visit id
```
Comment (WuTheFWasThat) [inline assets/js/data.coffee]: nit: put a newline between functions, in general

**Comment by WuTheFWasThat** [comment]
it looks mostly fine.  i kinda suspect these ancestor hooks aren't exactly what we want (for neither marks nor time-tracking).  need to think about it a bit more though

File: assets/js/data.coffee (hunk 2 above)
Comment (za3k) [inline assets/js/data.coffee]: What should be before what?

Comment (WuTheFWasThat) [inline assets/js/data.coffee]: as in you're emitting the wrong event :P

File: assets/js/data.coffee [hunk 4]
```diff
@@ -4,6 +4,7 @@ if module?
   global.utils = require('./utils.coffee')
   global.errors = require('./errors.coffee')
   global.constants = require('./constants.coffee')
+  global.EventEmitter = require('./eventEmitter.coffee')
```
Comment (WuTheFWasThat) [inline assets/js/data.coffee]: this was already added, rebase issue probably

Comment (za3k) [inline assets/js/data.coffee]: fixed

**Pull request closed without merging**